{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8eef668d-a4ab-4b59-bd55-08033f0048e6",
   "metadata": {},
   "source": [
    "# Statistical Analysis of RNAseq Data\n",
    "\n",
    "This notebook contains some python code that will collate the `kallisto` outputs you generated in the previous notebook and store these as a python object called a dataframe. If you are not familiar with python, you can conceptualise a dataframe as being a spreadsheet that python uses to store and retrive data. \n",
    "\n",
    "You don't need to write any code here, but if you are curious, I encourage you to look at the code I have written and see if you work out what is going on.\n",
    "\n",
    "After the dataframe has been made, there is some additional python code that passes this to different programs from software package called `pydeseq2`. These programs process the data and carry out a statistical test that compares transcript counts between the healthy and colorectal cancer patient groups to identify genes that are differentially expressed between the two groups. Finally, a table is printed that shows the names of the genes that were found to have statistically significant changes in experssion level between the two groups after adjusting for multiple comparisons.\n",
    "\n",
    "## Further reading for those of you who want to know more about how all of this works\n",
    "\n",
    "__________________\n",
    "\n",
    "**Publication describing `pydeseq2`:** \n",
    "\n",
    "https://academic.oup.com/bioinformatics/article/39/9/btad547/7260507\n",
    "\n",
    "___________________\n",
    "\n",
    "**Documentation for `pydeseq2`:**   \n",
    "\n",
    "https://pydeseq2.readthedocs.io/en/latest/\n",
    "\n",
    "___________________\n",
    "\n",
    "**Publication describing the statistical methods that underpin `pydeseq2`:** \n",
    "\n",
    "https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8\n",
    "\n",
    "__________________"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad365d96-beba-4a31-860e-b497a3a66b6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75f818cb-8d07-48c2-9b37-2e0af896d98f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd ~339_lab-main/outputs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff5c6b6a-a5ee-4515-8654-da8a8169009e",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_directory = './' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d5f2929-c3ee-4907-87a2-37962b461a71",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_tsv_files(start_directory):\n",
    "    tsv_files = []\n",
    "    for root, _, files in os.walk(start_directory):\n",
    "        for file in files:\n",
    "            if file.endswith('.tsv'):\n",
    "                full_path = os.path.join(root, file)\n",
    "                tsv_files.append(full_path)\n",
    "    return tsv_files\n",
    "\n",
    "def concatenate_tsv_files(tsv_files):\n",
    "    dfs = []\n",
    "    for file_path in tsv_files:\n",
    "        df = pd.read_csv(file_path, sep='\\t')\n",
    "        df['file_path'] = file_path\n",
    "        dfs.append(df)\n",
    "    concatenated_df = pd.concat(dfs, ignore_index=True)\n",
    "    return concatenated_df\n",
    "    \n",
    "def gather_and_concat(start_directory):\n",
    "    tsv_files = find_tsv_files(start_directory)\n",
    "    concatenated_df = concatenate_tsv_files(tsv_files)\n",
    "    return concatenated_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91c567c8-a921-4310-8dfb-c33f39cf0ce7",
   "metadata": {},
   "outputs": [],
   "source": [
    "concatenated_df = gather_and_concat(start_directory)\n",
    "\n",
    "#make an all zeros df with the uniqe patient and gene names\n",
    "row_names=concatenated_df['file_path'].unique().tolist()\n",
    "column_names=concatenated_df['target_id'].unique().tolist()\n",
    "count_df= pd.DataFrame(1, index=row_names, columns=column_names)\n",
    "\n",
    "#update the empty count df with the appropriate values by iterating the concatenated df\n",
    "# Initialize an empty list to store the tuples\n",
    "xyz_tuples = []\n",
    "\n",
    "# Iterate through each row in the DataFrame and get values from columns 'x', 'y', and 'z'\n",
    "for row in concatenated_df.itertuples(index=False):\n",
    "    x_value = getattr(row, 'file_path')\n",
    "    y_value = getattr(row, 'target_id')\n",
    "    z_value = getattr(row, 'tpm')\n",
    "    \n",
    "    # Create a tuple from the values and add it to the list\n",
    "    xyz_tuples.append((x_value, y_value, z_value))\n",
    "\n",
    "#use tuple  values to update count df\n",
    "for i in xyz_tuples:\n",
    "    count_df.loc[i[0]][i[1]]=int(i[2])\n",
    "\n",
    "count_df=count_df.sort_index()\n",
    "\n",
    "genes_to_keep = count_df.columns[count_df.sum(axis=0) >= 10]\n",
    "count_df = count_df[genes_to_keep]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3236aa01-2726-4e9c-8bf8-31a80585e17c",
   "metadata": {},
   "outputs": [],
   "source": [
    "groups = ['c','c','c','h','h','h','h','h','h','c','c','c']\n",
    "metadata = pd.DataFrame(groups, index=sorted(row_names), columns=[\"group\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e5a42d6-9f8f-4f4f-8542-cf243e0e539b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dds = DeseqDataSet(\n",
    "    counts=count_df,\n",
    "    metadata=metadata,\n",
    "    design_factors=\"group\",\n",
    "    refit_cooks=True,\n",
    "    n_cpus=1,)\n",
    "\n",
    "dds.deseq2()\n",
    "stat_res = DeseqStats(dds, n_cpus=1)\n",
    "stat_res_c_vs_h = DeseqStats(dds, contrast=[\"group\", \"c\", \"h\"],alpha=0.05, n_cpus=1,cooks_filter=True, independent_filter=True)\n",
    "stat_res_c_vs_h.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfe14fb5-896b-4316-ac86-100992c84b76",
   "metadata": {},
   "outputs": [],
   "source": [
    "stat_res_c_vs_h.results_df[stat_res_c_vs_h.results_df['padj']<0.05]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39f0bbb1-0f84-433c-b1f2-e146f46f78a7",
   "metadata": {},
   "source": [
    "# What can we learn from these results?\n",
    "\n",
    "If all has gone according to plan, the table above should contain gene names in the first row that end with `genome_9_orf00042`, `genome_9_orf00043`, `genome_9_orf00045` and  `genome_9_orf00046` All of the carryon before this is some weird side effect of the helper code I wrote, and I haven't figured out how to remedy this just yet. \n",
    "\n",
    "Essentially, this is saying that the genes **orf00042, orf00043, orf00045 and  orf00046** from **genome_9** are differentially expressed between the two groups we compared.\n",
    "\n",
    "If we look at the `log2FoldChange` column, all of the values are positive, which tells us they were **upregulated** in the colorectal cancer group\n",
    "\n",
    "All of the values in the `padj` column less than 0.0005, which tells us that after adjusting for multiple comparisons, these differences are **highly statistically significant**.\n",
    "\n",
    "**What should we do now?** \n",
    "\n",
    "Well .... bcause I made up these data, I know exactly what we should do now!\n",
    "\n",
    "In the final cell of this note book, I have added the gene sequences for orf00042, orf00043, orf00045 and  orf00046 from genome_9. We would like to see what these genes might be doing. A good starting point is to compare them to a sequence database to see if they match any genes with known function. We will use the online version of the sequence search tool **blastx** to do this.\n",
    "\n",
    "The link below will take you to the blastx webserver. Once you are there, you can cut and paste all of the sequences into the search box and click the **blast** button at the bottom of the page. The default parameters should be fine for this.\n",
    "\n",
    "**Link for blastx:**\n",
    "\n",
    "https://blast.ncbi.nlm.nih.gov/Blast.cgi?LINK_LOC=blasthome&PAGE_TYPE=BlastSearch&PROGRAM=blastx\n",
    "\n",
    "When the search has finished running, you will automatically be redirected to a page that displays the results. Using the drop down menu next to the line labelled `Results for` near the top of the page, you can select which of the sequences you want to view the results for.\n",
    "\n",
    "**And then what?** \n",
    "\n",
    "That's up to you! Here are some suggestions to get you started: \n",
    "\n",
    " - Look at the names of top few hits for each of the genes\n",
    " \n",
    " - Do they share anything in common?\n",
    " \n",
    " - Click on the links for some of these and see if you can find the publication(s) that relate to the hit sequence(s)\n",
    " \n",
    " - What can you learn from these publications?\n",
    " \n",
    " - Do the known functions of these genes suggest anything about the link between the microbiome and cancer risk. Remember, we know that the genes we searched were significantly upregulated in the microbiomes of patients who went on to develop colorectal cancer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d512413-a236-4fb3-a0b1-bd3148c6ce73",
   "metadata": {},
   "source": [
    "```\n",
    ">genome_9_orf00042 \n",
    "ATGACGAAGGATGTCGCACTGATGTTCCCTGGCTCCGGTTCGCAATATGTAGGCATGGCA\n",
    "CGGTGGCTGTATGAGCGTTATCCGCAGGTGCGTACTCTGTTTGATGAAGCTAGCCAGATC\n",
    "ACGGAACGGGACATGGCAGCGCTGTGCCTGTCGGGTACTTTGGTACAACTTGCTGAGCCG\n",
    "ACGGCAATGGCGCTGGCTATCTATACCACCAGCGTGGCCCACTTTGTCGCGTGGCAGCAG\n",
    "TTTTTGGCGCAAACCGGTGTCCATGTCAACCTACGTTATATGTTGGGTCATAGTTTGGGC\n",
    "GAATATGCTGCGCTGACCTGTAGTGGTGCGCTGAGCTTTTCCCAAGCACTGGCGCTGGTG\n",
    "GCGATGCGTAGCCGTTTAGCCAGTGAGATCGCCCGCGAAATGGACGCGAGTACCACTATC\n",
    "ATCAAGCAAGGGAATCAGGCACTGGTGGCAGCGGCGTGCGAGGTGGCGGAGCGTGAAACC\n",
    "CGACAGCAGGTCGGTATTGCCTGCTTCAATTCGCCGCAACAGTTTATGTTATCCGGACAA\n",
    "AACTCGGCCATTATCGCCGCCGAGCAATACCTGCTGGATCATGATCGCCAAGTCGAAGTA\n",
    "GTGCCGCTGATTGGTGGTGTTCCTTACCACAGCCCACTATTGAAACCTTGCGGACAGCAA\n",
    "TTGCGCAAGGCGCTGGATCGCTGCGAATGGCGACGTCCGTGCTGCCCAGTGATCTCCAAT\n",
    "GTCAACGCTCAACCCTATCCCGATACGGTCACGCCAGTCCAGTGGCTAGAACAACAGCTC\n",
    "TCTCAGCCAGTGCAGTGGCAGCGCTCTCTCACCTACCTGACAGGACACTTGTCACCGATC\n",
    "GCGATAGAGATCGGTCCGCAAAGTGTGCTGAAAAACCTGCTGCTGGAGAATCGCTATCCA\n",
    "GCACCAGTTTACGCATTTGACAATCGTCATGATCGTGCGCAACTAGCATTGGTGCTGGGG\n",
    "GATAACATGGCGGTGAAGACCGATCCGGAAGCCGTGCGCCGCCAACGCATCACGCTGCTC\n",
    "ACTAACGCCCTGACAGCCACACGCCATCACCGCGCCGCCGATGTGGCTGCCAGTGCGCAA\n",
    "TTGAAAGAATTGCTCAGTCGTTTTTTTGAGCGCATCCAGCAGATTGAGCAGCGCGGCACA\n",
    "TCCAGCGAAGAGGACATTGCTTTTTTGCATGAGCTTCTTGAACAGGGATTTCAACTCAAA\n",
    "GGCAGTAGTCAGGCAGAAATCGACGCCTGCCACGCGCGGTTGGCTTCCGACAACGGCGGA\n",
    "CAGGCG\n",
    ">genome_9_orf00043\n",
    "ATGTGCACTGAAAATTATGAGCTGGCTCAGCAAGAGGCAGTCCTGTTTGCCAAACAACAT\n",
    "CTCGCCCTGGCTGCACAGAATATTGAACGTCAGCAGTTTATTGTGCCAGACATTATTTCT\n",
    "TGCGTGGCGCAGGCCGGGTACTTAGGTGCGTCAATCCCCCAAAAATATGGCGGACGAGGT\n",
    "TACGATTCTTATCAACTGTGCGCTTTGCATGAAGTGATGGCTGGGGTTCACGGTTCGTTA\n",
    "GAAAATCTCATAACAGTGACTGGCATGGTCAGCACGCTGCTGCAACGCGTGGGCAGCGCA\n",
    "GCGCAAAAAGCCCATTATTTGCCAAAGCTGGCAACGGGGGAATTGATCGGCGCGATTGCC\n",
    "CTCACAGAGCCGAATATCGGCAGTGACTTAGTGAACGTGGAAACAGAATTACAGCAGGAT\n",
    "GGCGACGGTTGGCGGCTGAACGGGAAAAAGAAATGGATCACGCTAGGGCAGATTGCTGAT\n",
    "TTTTTTATTGTTTTAATCCACTGTGGTAATCAATTAGCGACAGTGCTTATCGATCGCAAT\n",
    "ACCGACGGCTTTACTATCACTCCACTCAACGACATGTTGGGGTTACGCGGCAATATGTTG\n",
    "GCAGAGCTGCATTTTAATGACTGCCGCCTGAAGGAGGATGCATTGCTCGGTCCGCTCACC\n",
    "CCTGGGGTACCGCTGGCGGTGAATTTTGCGCTCAATGAAGGGCGGTTCACCACCGCTTGT\n",
    "GGCAGCCTAGGATTATGTCGGGCCGCTGTGGATGTGGCAGCGCGCTACATCCGCCAGCGT\n",
    "AAACAGTTCAAGCGGCGGTTATTTTCTCATGGCATTGTTCAGCACTTGTTTGCCACAATG\n",
    "CTGACCCAAACGCGCAGTGCGCAACTAATGTGCTTCAGCGCGGCGGAATACCGCGAAACG\n",
    "CTGCACCCGGCCATGATAAACCAAATCCTGATGGCGAAGTATGTCGCTTCCAAAGCTGCG\n",
    "GTGGACGTGGCCGGGAAGGCGGTGCAACTGCTCGGTGCGAATGGTTGCCATGCCGATTAT\n",
    "GCCGTCGAACGTTACTACCGCGACGCCAAAATCATGGAAATCATCGAAGGGACATCGCAA\n",
    "ATTCATGAAATCCAAATTGCGATGAATTACATGATGGGGAGTGAAGCA\n",
    ">genome_9_orf00045\n",
    "ATGAATATGAAAAAGCAAGATATGAAAGCCGCCATTCGGGAATTTCTTTCACGCTCATTA\n",
    "CGTGGGCATACGTTGAACGATGATGACGATATTTTTTCTCTCGGGCTTGTCCATTCGTTA\n",
    "TTTACTGTGCAAATCATACTGTTTATAGAAAAAAATTTTCAGGTTGAGCTGGAAGTGAGT\n",
    "GAGTTGAAAACAGAACAGATTGCTACCGTCAATAAAATAGTGGAGCTCATTCAGCGACAA\n",
    "ACAGGCCTGGAG\n",
    ">genome_9_orf00046\n",
    "ATGATGAACGTGGCGGTAATAGGTGCAGGAGTAATGGGAACTGGCGTCGCTCATAACATG\n",
    "GCGCAATACGGCATATCAACTAACGTTGTGGATATTTCTCAATCTCAGTTGGATAAATGC\n",
    "CGACAGATGATCGAAGCGAACTTACGCTTATATAATTTTCACCCGCAGCATAAAAAAAAG\n",
    "ACGCATAGCACAGCGGAGATAATGGAAAATATCCGTTTCACAACTGAGTTGGACGACATT\n",
    "GTAGAATGTGATTTGGTGATTGAGAATATTACTGAGGATATAGAGAAAAAAAATGCACTA\n",
    "TACACACGGATGAATACGATCTGCGGCGCGTCGACAGTGTTTGGTGTTAATACGTCCGCC\n",
    "ATATCTATCACTGCGTTGTCAAAATTAATGCGACATCCGGAGAATGTAGTCGGCGTCCAT\n",
    "TTTATGAACCCGGTACCGCTGATGCACACCGTCGAATTAATCCGTGGTGTACATACTGCA\n",
    "GAGCGTACACTGAACATATTTCACCACTTATTTGCTCAGTTGAATAAAACCGGCATCGTG\n",
    "GTCAATGATTCTCCGGGCTTTGTCACTAATCGTGCCATGATGATTTTTGTTAACGAAGCC\n",
    "ATATTTATGGTGCAGGAGCAGATCGCCCGTGCTGAGGATATCGATACGTTGTTTAAAACC\n",
    "TGCTTCGGGCACAAGATGGGGCCACTACAAACTGCCGATTTGATTGGCTTGGATACGATT\n",
    "TTGCAATCGTTGCAGGTGCTATATGAAAGTTTTAATGATGATAAGTATCGCCCCAGCTTT\n",
    "TTATTAAAGAAAATGGTCGATGCAGGGTATTTGGGCGTGAAATCAGGGCAGGGGTTTTAT\n",
    "CGCTATCAACAGACGTACGCCGAGCAG\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
